enigma-swing/src/main/java/cuchaz/enigma/gui/node/SortedMutableTreeNode.java

```diff
@@ -0,0 +1,108 @@
+package cuchaz.enigma.gui.node;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.List;
+
+import javax.swing.tree.DefaultMutableTreeNode;
+import javax.swing.tree.MutableTreeNode;
+import javax.swing.tree.TreeNode;
+
+import com.google.common.collect.Iterables;
+
+/**
+ * A MutableTreeNode whose contents are always guaranteed to be sorted with the given comparator.
+ */
+public class SortedMutableTreeNode extends DefaultMutableTreeNode {
+	private final Comparator<TreeNode> comparator;
+	private final List<TreeNode> children;
+	private boolean isSorted = true;
+
+	public SortedMutableTreeNode(Comparator<TreeNode> comparator) {
+		this.comparator = comparator;
+		this.children = new ArrayList<>();
+	}
+
+	@Override
+	public void insert(MutableTreeNode child, int index) {
+		if (child == null) {
+			throw new IllegalArgumentException("child is null");
+		}
+
+		MutableTreeNode oldParent = (MutableTreeNode) child.getParent();
+
+		if (oldParent != null) {
+			oldParent.remove(child);
+		}
+
+		child.setParent(this);
+		this.children.add(child);
+		this.isSorted = false;
+	}
+
+	private void checkSorted() {
+		if (!this.isSorted) {
+			this.isSorted = true;
+			this.children.sort(this.comparator);
+		}
+	}
+
+	@Override
+	public void remove(int index) {
+		checkSorted();
+
+		remove((MutableTreeNode) getChildAt(index));
+	}
+
+	@Override
+	public void remove(MutableTreeNode node) {
+		this.children.remove(node);
+		node.setParent(null);
+	}
+
+	@Override
+	public TreeNode getChildAt(int childIndex) {
+		checkSorted();
+
+		return this.children.get(childIndex);
+	}
+
+	@Override
+	public int getChildCount() {
+		return this.children.size();
+	}
+
+	@Override
+	public int getIndex(TreeNode node) {
+		return Iterables.indexOf(this.children, other -> this.comparator.compare(node, other) == 0);
```

Comment:
Shouldn't this call `checkSorted()` as well, since the order of the node in the list before and after sorting may be different? (Or, if this is intended, perhaps a comment is in order.)

Also, why use `Iterables.indexOf` instead of `List.indexOf`? If the comparator is known to be consistent with `equals` (and the javadoc for `Comparator` recommends that comparators used with sorted collections be consistent with equals), then both ways should be equivalent. (Theoretically, `List.indexOf` would be faster by virtue of the `List` implementation taking advantage of impl. details.)

Reply:
Can't the `ClassEntry` comparator be inconsistent with `equals`, though?
In any case, I'm probably just going to turn this into a manual for loop.

Reply:
As far as I can tell, the `ClassEntry` comparators -- defined in [`ClassSelector#DEOBF_CLASS_COMPARATOR`](https://github.com/FabricMC/Enigma/blob/a49ff1b756fa64b4b1f10b975fd1dd66abb00f87/enigma-swing/src/main/java/cuchaz/enigma/gui/ClassSelector.java#L41) and used in [`DeobfPanel#<init>`](https://github.com/BasiqueEvangelist/Enigma/blob/e0740498ee3029e6461febc04a5d5163cc590be3/enigma-swing/src/main/java/cuchaz/enigma/gui/panels/DeobfPanel.java#L28), and defined as a lambda in [`ObfPanel#<init>`](https://github.com/BasiqueEvangelist/Enigma/blob/e0740498ee3029e6461febc04a5d5163cc590be3/enigma-swing/src/main/java/cuchaz/enigma/gui/panels/ObfPanel.java#L24-L33) -- are both consistent with `equals`.

`ClassEntry#equals` checks if the parent and class name of the entry are equivalent; the comparators above compare using the full name of the class[^1], which is derived from the class name of the entry and of its parent. Because of that, if two `ClassEntry`s are equal in terms of `equals`, the comparators will also declare them equivalent (returning `0`).

Though, that is the present state of the code; the future may possibly bring comparators inconsistent with `equals`, but I think that's improbable (and would already cause problems anyway, because then you'd have entries that don't equal themselves or are equal to other entries according to the comparator).

[^1]: In the `ObfPanel` comparator's case which compares the lengths of the full names first: if the two full names are equivalent, then necessarily the lengths are also the same.